### ISSUE TITLE
Regarding Saving checkpoints

### ISSUE THREAD

**Issue by wsskuleuven** [MIC-DKFZ/nnUNet#1286]
Dear Fabian,

I am confusing about how does the checkpoint-saving process works. In the description it says "nnU-Net stores a checkpoint every 50 epochs". However, with the nnUNetTrainerV2, it looks like each checkpoint is saved after each epoch (An example is attached at the end).

So if it saves checkpoint after each epoch, I think when I stop the training manually at e.g., the 499th epoch, I can get the current model (i.e., model trained after 499 epochs), instead of model trained with 450 epochs (i.e., saved after 50*N epochs). Is that correct?

Thanks in advance.

2022-12-13 13:33:40.737566: 
epoch:  15 
2022-12-13 13:35:28.720252: train loss : -0.3791 
2022-12-13 13:35:38.465125: validation loss: -0.3852 
2022-12-13 13:35:38.467820: Average global foreground Dice: [0.9008, 0.4931] 
2022-12-13 13:35:38.469846: (interpret this as an estimate for the Dice of the different classes. This is not exact.) 
2022-12-13 13:35:39.061119: lr: 0.009897 
2022-12-13 13:35:39.181365: **saving checkpoint...** 
2022-12-13 13:35:40.390207: done, saving took 1.33 seconds 
2022-12-13 13:35:40.448827: This epoch took 119.709329 s
 
2022-12-13 13:35:40.450899: 
epoch:  16 
2022-12-13 13:37:34.267906: train loss : -0.3850 
2022-12-13 13:37:45.648075: validation loss: -0.4097 
2022-12-13 13:37:45.650818: Average global foreground Dice: [0.9096, 0.5315] 
2022-12-13 13:37:45.653142: (interpret this as an estimate for the Dice of the different classes. This is not exact.) 
2022-12-13 13:37:46.212921: lr: 0.009891 
2022-12-13 13:37:46.321118: **saving checkpoint...** 
2022-12-13 13:37:47.084156: done, saving took 0.87 seconds 
2022-12-13 13:37:47.104152: This epoch took 126.650896 s

**Comment by JackRio** [comment]
```
2023-01-24 21:13:43.991983: lr: 0.005359
2023-01-24 21:13:44.000225: saving scheduled checkpoint file...
2023-01-24 21:13:44.153094: saving checkpoint...
2023-01-24 21:13:44.927667: done, saving took 0.93 seconds
2023-01-24 21:13:45.046274: done
2023-01-24 21:13:45.052797: This epoch took 116.880993 s
```

This is how the log should look like if its doing a scheduled checkpoint after every 50 epoch default.

https://github.com/MIC-DKFZ/nnUNet/blob/aa53b3b87130ad78f0a28e6169a83215d708d659/nnunet/training/network_training/network_trainer.py#L577

The saving you see triggers because the current epoch has best moving average on the validation criterion.

**Comment by wsskuleuven** [comment]
> ```
> 2023-01-24 21:13:43.991983: lr: 0.005359
> 2023-01-24 21:13:44.000225: saving scheduled checkpoint file...
> 2023-01-24 21:13:44.153094: saving checkpoint...
> 2023-01-24 21:13:44.927667: done, saving took 0.93 seconds
> 2023-01-24 21:13:45.046274: done
> 2023-01-24 21:13:45.052797: This epoch took 116.880993 s
> ```
> 
> This is how the log should look like if its doing a scheduled checkpoint after every 50 epoch default.
> 
> https://github.com/MIC-DKFZ/nnUNet/blob/aa53b3b87130ad78f0a28e6169a83215d708d659/nnunet/training/network_training/network_trainer.py#L577
> 
> The saving you see triggers because the current epoch has best moving average on the validation criterion.

Thank you JackRio! Your answer perfectly solved my question.

**Comment by FabianIsensee** [comment]
Correct. We save every 50 (checkpoint_latest) and again every time we encounter a new best epoch (checkpoint_best). Thanks again @JackRio
